rust/tw_evm/src/abi/function.rs

```diff
@@ -0,0 +1,87 @@
+// Copyright © 2017-2023 Trust Wallet.
+//
+// This file is part of Trust. The full Trust copyright notice, including
+// terms governing use, modification, and redistribution, is contained in the
+// file LICENSE at the root of the source code distribution tree.
+
+use crate::abi::decode::decode_params;
+use crate::abi::param::Param;
+use crate::abi::param_token::NamedToken;
+use crate::abi::token::Token;
+use crate::abi::{AbiError, AbiErrorKind, AbiResult};
+use serde::Deserialize;
+use tw_memory::Data;
+
+#[derive(Clone, Debug, Default, Deserialize)]
+pub struct Function {
+    /// Function name.
+    pub name: String,
+    /// Function input.
+    pub inputs: Vec<Param>,
+    /// Function output.
+    pub outputs: Vec<Param>,
+    /// Whether the function reads or modifies blockchain state.
+    #[serde(rename = "stateMutability", default)]
+    pub state_mutability: ethabi::StateMutability,
+}
+
+impl Function {
+    /// Returns a signature that uniquely identifies this function including `inputs` **only**.
+    ///
+    /// Examples:
+    /// - `functionName()`
+    /// - `functionName()`
+    /// - `functionName(bool)`
+    /// - `functionName(uint256,bytes32)`
+    ///
+    /// The method implementation is inspired by
+    /// https://github.com/rust-ethereum/ethabi/blob/b1710adc18f5b771d2d2519c87248b1ba9430778/ethabi/src/function.rs#L88-L97
+    pub fn signature_with_inputs(&self) -> String {
+        let inputs = self
+            .inputs
+            .iter()
+            .map(|p| p.kind.to_ethabi().to_string())
+            .collect::<Vec<_>>()
+            .join(",");
+        format!("{}({inputs})", self.name)
+    }
+
+    /// Parses the ABI function input to a list of tokens.
+    pub fn decode_input(&self, data: &[u8]) -> AbiResult<Vec<NamedToken>> {
+        decode_params(&self.inputs, data)
+    }
+
+    /// Encodes function input to Eth ABI binary.
+    pub fn encode_input<I>(&self, tokens: I) -> AbiResult<Data>
+    where
+        I: IntoIterator<Item = Token>,
+    {
+        let ethabi_tokens: Vec<_> = tokens.into_iter().map(Token::into_ethabi_token).collect();
+        self.to_ethabi_function()
+            .encode_input(&ethabi_tokens)
+            // Given tokens don't match the ABI.
+            .map_err(|_| AbiError(AbiErrorKind::Error_abi_mismatch))
+    }
+
+    pub(crate) fn to_ethabi_function(&self) -> ethabi::Function {
+        let inputs = self
+            .inputs
+            .iter()
+            .map(Param::to_ethabi_param)
+            .collect::<Vec<_>>();
+        let outputs = self
+            .outputs
+            .iter()
+            .map(Param::to_ethabi_param)
+            .collect::<Vec<_>>();
+
+        #[allow(deprecated)]
```

Comment:
Why are we using a deprecated function here?

Reply:
The function has deprecated `constant` param, but we have to specify it on the object creation.
```rust
#[deprecated(note = "The constant attribute was removed in Solidity 0.5.0 and has been \
                        replaced with stateMutability.")]
/// Constant function.
#[cfg_attr(feature = "serde", serde(skip_serializing_if = "Option::is_none"))]
pub constant: Option<bool>,
```

Even in ethabi tests they use `#[allow(deprecated)]`: [link](https://github.com/rust-ethereum/ethabi/blob/b1710adc18f5b771d2d2519c87248b1ba9430778/ethabi/src/function.rs#L110-L120)

Usually, there is `default` function, so the deprecated fields can be skipped, but not in that case:
```rust
		let func = Function {
			name: "baz".to_owned(),
			inputs: vec![
				Param { name: "a".to_owned(), kind: ParamType::Uint(32), internal_type: None },
				Param { name: "b".to_owned(), kind: ParamType::Bool, internal_type: None },
			],
			outputs: vec![],
			state_mutability: StateMutability::Payable,
		        ..Function::default()
		};
```